keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we propose defaulting failure policy to "fail" and allowing it to
+be configured on a per-policy basis:
+
+```yaml
+  failurePolicy: Ignore # The default is "Fail"
+  validations:
+    - expression: "object.spec.xyz == configuration.x"
+      
+```
+
+<<[UNRESOLVED jpbetz ]>>
+Should failurePolicy be per-validation expression or per-policy?
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED jpbetz ]>>
+Also allow for a `failurePolicy` field on policy configuration resources?
+<<[/UNRESOLVED]>>
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
+- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
+  `MutatingWebhookConfiguration`.
+
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
+
+Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+#### Singleton policies
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+Should this be supported in phase 1? Can it be deferred to phase 2 without
+necessitating breaking API changes? What are the implications on how violations
+are reported.
+<<[/UNRESOLVED]>>
+
+For a policy that requires no configuration, we would prefer not to ask users
+create configuration CRD and configuration resource that serve no actual
+purpose. Instead, we will allow a `ValidatingAdmissionPolicy` to define
+a singleton policy, the recipe is:
+
+- Set a `policyType: Singleton` field
+- Exclude `spec.config`
+
+For example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  policyType: Singleton
+  match:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+  - expression: "object.spec.replicas < 100"
+```
+
+Adding a `policyType` field makes it easier for a user examining the policy to
+observe that the policy is different that normal (configurable) policies. This
+is much less obvious if we infer that a policy is a singleton by the state of
```

Comment:
not so sure. If I don't fill in config and I don't reference any config, it seems pretty straightforward.